### ISSUE TITLE
[Bug]: Milvus HNSW no reproducibility or How to set random_seed for HNSW

### ISSUE THREAD

**Issue by weilong-web** [milvus-io/milvus#25908]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.2.1
- Deployment mode(standalone or cluster):cluster
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): Ubuntu
- CPU/Memory: 2TB
- GPU: A100 80GB
- Others:
```

### Current Behavior

Setting the random seed does not seem to have the desired effect, as I am observing different retrievers each time I run the code, even with identical parameters and data. I'm using HNSW. Instead, FLAT is fine.

### Expected Behavior

When I set random seed like below:
```
		seed = 42
		random.seed(seed)
		os.environ['PYTHONHASHSEED'] = str(seed)
		np.random.seed(seed)
		torch.manual_seed(seed)
		torch.cuda.manual_seed(seed)
		torch.use_deterministic_algorithms(True)
		torch.cuda.manual_seed_all(seed) # if you are using multi-GPU.
		torch.backends.cudnn.benchmark = False
		torch.backends.cudnn.deterministic = True
		enable_full_determinism(seed)
		set_seed(seed)
		tf.random.set_seed(seed)
		tf.config.experimental.enable_op_determinism()	
```
Each time it should output same results.

### Steps To Reproduce

```markdown
1. Use HNSW.
2. Set seed.
3. Output the question results.
```

**Comment by yanliang567** [comment]
/assign @liliu-z 
any comments?

/unassign

**Comment by liliu-z** [comment]
Sorry I didn't fully get your issue. Did you mean that you can not get the same results with the same query?
Can I ask what index you are using? CPU-IVFFLAT or HNSW?
And about the seed, can I ask where did you set the seed?

**Comment by weilong-web** [comment]
Thank you for your answer.
Yes, I can not get the same results with the same query.
My problem is for HNSW.
```
from langchain.vectorstores import Milvus
from pymilvus import connections, Collection
import torch
import random
import numpy as np
import os
from transformers import enable_full_determinism, set_seed
import tensorflow as tf
MILVUS_HOST = *****"; MILVUS_PORT = "*****"
MILVUS_CONNECTION_ARGS = {"host": MILVUS_HOST, "port": MILVUS_PORT}
INDEX_PARAMS = {"index_type": "IVF_FLAT", "metric_type": "L2", "params":{"nprobe": 10}}

class Vectorstore:
	""" Essential wrapper functions for using the milvus vectorstore
	TODO: This is a double-wrap (langchain + nerve) When confident, make this the default vectorstore
	"""

	def __init__(self, args):
		self.args = args
		seed = 42
		random.seed(seed)
		os.environ['PYTHONHASHSEED'] = str(seed)
		np.random.seed(seed)
		torch.manual_seed(seed)
		torch.cuda.manual_seed(seed)
		torch.use_deterministic_algorithms(True)
		torch.cuda.manual_seed_all(seed) # if you are using multi-GPU.
		torch.backends.cudnn.benchmark = False
		torch.backends.cudnn.deterministic = True
		enable_full_determinism(seed)
		set_seed(seed)
		tf.random.set_seed(seed)
		tf.config.experimental.enable_op_determinism()			
	def get_collection_name(self):
		""" Uses job args to create a valid milvus collection name
		NOTE: milvus collection name can only contain letters, numbers, and underscore
		"""
		args = self.args
		return args.dataset.name.replace("/", "_").replace(".", "_")
	
	def get_collection(self):
		connections.connect("default", host=MILVUS_HOST, port=MILVUS_PORT)
		collection = Collection(self.get_collection_name())
		return collection	
	
	def get_retriever(self, encoder):
		""" Returns a langchain retriever for the milvus collection
		"""
		args = self.args
		db = Milvus(
		   encoder,
		   collection_name=self.get_collection_name(),
		   connection_args=MILVUS_CONNECTION_ARGS)
		retriever = db.as_retriever()
		return retriever
```

**Comment by liliu-z** [comment]
@weilong-web Sorry, looks like you are using langchain and I am still a little bit confused about how you are working with it. Plz CMIIW, you were trying to get multiple retrievers by call this `get_retriever()` multiple times for a same `Vectorstore` instance? Then you noticed the results are different from retrievers? BTW, if yes, what's the relationship with this problem and random seed?
/assign @filip-halt

**Comment by weilong-web** [comment]
You are correct. I provide the same instance each time, the retrievers can differ. I guess it is due to the inherent randomness and non-determinism in HNSW. 
I have attempted to address this by setting a random seed, but unfortunately, it did not yield the desired reproducibility. To maintain consistency and predictability in the retrievers, I seek guidance on effectively setting the random seed or any other method for HNSW in order to avoid randomness and obtain reliable results.

**Comment by liliu-z** [comment]
> You are correct. I provide the same instance each time, the retrievers can differ. I guess it is due to the inherent randomness and non-determinism in HNSW. I have attempted to address this by setting a random seed, but unfortunately, it did not yield the desired reproducibility. To maintain consistency and predictability in the retrievers, I seek guidance on effectively setting the random seed or any other method for HNSW in order to avoid randomness and obtain reliable results. @liliu-z @filip-halt

Are you using Milvus 2.2.1? To my understanding HNSW's search is reproducible. Plz let me know if there is any random factors in langchain side @filip-halt

**Comment by weilong-web** [comment]
I write a simple program to do test the search process using HNSW, it is reproducible.
Now the problem should be from Langchain `RetrievalQA`, since I call this function for decoding process. Any suggestion on reproducibility for `RetrievalQA` @filip-halt 

```
from langchain.chains import RetrievalQA

gar_chain = RetrievalQA.from_chain_type(llm=self.decoder, chain_type="stuff", retriever=retriever, return_source_documents=True)
```

**Comment by filip-halt** [comment]
What do you results look like, what is different about them? Anything Milvus related should not be affected by the seed. Why is a different seed constantly being set? This might affect things earlier in the pipeline
